### PR TITLE
Allow snapshot publishing and add ShouldPublish parameter

### PR DESCRIPTION
**Pull request by hallipr**
Because we want ShouldPublish to conditionally include the java-publishing.yml steps and parameters and static variables can be used in template conditionals, we need ShouldPublish to be a parameter, not a runtime variable.  To avoid having both variables and parameters in the pipeline, BlobPath and StageOnly are also converted from variables to parameters.

This improves the UI experience, but will prevent us from creating multiple pipelines over the same yaml with different default values because Azure DevOps only allows for default variables in a pipeline definition, not default parameter values.

### CODE REVIEW

**Comment by hallipr** [comment]
resolves: #13406

File: eng/scripts/Publish-MavenPackages.ps1 [hunk 1]
```diff
@@ -145,6 +145,17 @@ foreach ($packageDetail in $packageDetails) {
   }
   elseif ($RepositoryUrl -like "https://oss.sonatype.org/service/local/staging/deploy/maven2/") {
     if (!$StageOnly) {
```
Comment (hallipr) [inline eng/scripts/Publish-MavenPackages.ps1:147]: This logic isn't right...

If StageOnly, we don't want to release a snapshot.  Staging snapshots doesn't appear to be supported.

File: eng/scripts/Publish-MavenPackages.ps1 [hunk 2]
```diff
@@ -48,13 +48,19 @@ if ($packageDetails.Length -eq 0) {
...
+      throw "Package $($packageDetail.FullyQualifiedName) is a Snapshot and StageOnly is set to 'true'. Staging of snapshot packages is not supported."
+    }
+  }
+}
+
 Write-Host "Starting GPG signing and publishing"
 
 foreach ($packageDetail in $packageDetails) {
   Write-Host "GPG signing and publishing package: $($packageDetail.FullyQualifiedName)"
-  $localRepositoryDirectory = Get-RandomRepositoryDirectory
```
Comment (hallipr) [inline eng/scripts/Publish-MavenPackages.ps1:55]: All of the options that only apply to non-snapshot maven central packages are moved to the maven central block

File: eng/scripts/Publish-MavenPackages.ps1 [hunk 3]
```diff
@@ -48,13 +48,19 @@ if ($packageDetails.Length -eq 0) {
   throw "Aborting, no packages to publish."
 }
 
+if($StageOnly)
```
Comment (weshaggard) [inline eng/scripts/Publish-MavenPackages.ps1:51]: nit:
```suggestion
if ($StageOnly)
```

File: eng/pipelines/partner-release.yml [hunk 4]
```diff
@@ -36,11 +50,12 @@ steps:
...
-- template: tools/gpg/gpg.yml@azure-sdk-build-tools
-- template: /eng/pipelines/templates/steps/java-publishing.yml
-  parameters:
-    ArtifactDirectory: $(Artifacts)
-    Target: MavenCentral
-    BuildToolsPath: $(Pipeline.Workspace)/azure-sdk-build-tools
-    JavaRepoRoot: $(Pipeline.Workspace)/azure-sdk-for-java
-    StageOnly: $(StageOnly)
+- ${{ if eq(parameters.ShouldPublish, 'true') }}:
+  - template: tools/gpg/gpg.yml@azure-sdk-build-tools
```
Comment (weshaggard) [inline eng/pipelines/partner-release.yml:54]: In a follow-up we should try to do the gpg signing all the time and upload those as artifacts so we have the exact files we plan to upload.

**Review by weshaggard — APPROVED** [review]

File: eng/scripts/Publish-MavenPackages.ps1 (hunk 3 above)
Comment (hallipr) [reply eng/scripts/Publish-MavenPackages.ps1:51]: Will address in the GPG artifact pr

**Pull request merged**